hudi-client/src/main/java/org/apache/hudi/table/action/commit/UpsertPartitioner.java

```diff
@@ -301,7 +301,7 @@ protected static long averageBytesPerRecord(HoodieTimeline commitTimeline, int d
               .fromBytes(commitTimeline.getInstantDetails(instant).get(), HoodieCommitMetadata.class);
           long totalBytesWritten = commitMetadata.fetchTotalBytesWritten();
           long totalRecordsWritten = commitMetadata.fetchTotalRecordsWritten();
-          if (totalBytesWritten > 0 && totalRecordsWritten > 0) {
+          if (totalBytesWritten > hoodieWriteConfig.getParquetSmallFileLimit() && totalRecordsWritten > 0) {
```

Comment:
So the main change here is that the code will keep looking for a commit that wrote at least smallFileLimit bytes do we get as better value? Wonder if this will have any side effects. For eg if there are a bunch of small commutes like you are referring to, then it means we may fallback to the de fault vale lot more? 

What do think about introducing another config here that controls this threshold as a fraction of parquetSmallFileLimit? I feel we can destroy it to 0.5, as opposed to 1.0 that you have now..

Reply:
This bug would happen when a small commit made in a new partition. If we make a small commit to an existing partition, it will very likely be merged to the existing file so `totalBytesWritten` is should be a normal size. If there is at least one file in the timeline larger than a small file then it should be not using the default value.
I feel like adding a new config here would increase the complexity and could be difficult for the user to understand. Should we do that for this edge case?

Reply:
I understand the issues that your are solving. It’s real and we are on the same page there

> If there is at least one file in the timeline larger than a small file then it should be not using the default value

Actually condition is more like - at least one commit with total writes greater than small file. Let’s understand how this may play out with an example.

Let’s say the table is empty and we start doing small commits with 5 records each.

Commit 1 : no candidates for small files. inserts will get grouped into  (100k default) record buckets and written to file1_c1
Commit 2 : prev file is a small file; inserts turned to updates to file1_c2, but since last commit wrote only 5 records, it will use default value
...
Keeps happening until file1_cn when eventually rewritten to greater than small file. 
Commit n+1: will use derived record size instead of default value. But there are no more small files. A new file group file 2_cn+1 is created
Commit n+2: we will keep using commit n’s derived value as long as it’s on the active timeline..

A corner case here is : if commit n is archived before we have another commit that got file2 expand beyond small file size, e we will fall back to default value again - after We read all the commits in the active timeline.. This is why I suggest we have a config record.size.estimation.threshold to control what fraction of small parquet file size we can use for estimation.. 

Makes sense? (Alternatively we could look into somehow computing the bloom filter overhead based on configs and subtract that from totalBytesWritten.. it’s probably better? )

Reply:
Got your point. Making a small commit is already an edge case for me so I didn't think of continuously making small commits. But agree we can use a new config to have the flexibility.

We have discussed subtracting the bloom filter size before, but prefer to not doing it, IIUC. Major reasons:
> * Even we deduct the size of the bloom filter, there will be other metadata and the `totalWriteBytes` is still not representing the total record size. When the situation we discussed above happens, it is possible that the small files will still be produced.
> * This will increase the complexity when we handle other indexing like HbaseIndexing.